reporter/collector_reporter.go

```diff
@@ -0,0 +1,559 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package reporter // import "go.opentelemetry.io/ebpf-profiler/reporter"
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"maps"
+	"os"
+	"slices"
+	"time"
+
+	lru "github.com/elastic/go-freelru"
+	log "github.com/sirupsen/logrus"
+	"go.opentelemetry.io/collector/consumer/consumerprofiles"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pprofile"
+	semconv "go.opentelemetry.io/otel/semconv/v1.25.0"
+
+	"go.opentelemetry.io/ebpf-profiler/libpf"
+	"go.opentelemetry.io/ebpf-profiler/libpf/xsync"
+)
+
+// Assert that we implement the full Reporter interface.
+var _ Reporter = (*CollectorReporter)(nil)
+
+// OTLPReporter receives and transforms information to be Collector Collector compliant.
+type CollectorReporter struct {
+	cfg          *Config
+	nextConsumer consumerprofiles.Profiles
+
+	// stopSignal is the stop signal for shutting down all background tasks.
+	stopSignal chan struct{}
+
+	// hostmetadata stores metadata that is sent out with every request.
+	hostmetadata *lru.SyncedLRU[string, string]
+
+	// executables stores metadata for executables.
+	executables *lru.SyncedLRU[libpf.FileID, execInfo]
+
+	// cgroupv2ID caches PID to container ID information for cgroupv2 containers.
+	cgroupv2ID *lru.SyncedLRU[libpf.PID, string]
+
+	// frames maps frame information to its source location.
+	frames *lru.SyncedLRU[libpf.FileID, *xsync.RWMutex[map[libpf.AddressOrLineno]sourceInfo]]
+
+	// traceEvents stores reported trace events (trace metadata with frames and counts)
+	traceEvents xsync.RWMutex[map[traceAndMetaKey]*traceEvents]
+
+	// samplesPerSecond is the number of samples per second.
+	samplesPerSecond int
+}
+
+// NewCollector builds a new CollectorReporter
+func NewCollector(cfg *Config, nextConsumer consumerprofiles.Profiles) (*CollectorReporter, error) {
+	executables, err :=
+		lru.NewSynced[libpf.FileID, execInfo](cfg.ExecutablesCacheElements, libpf.FileID.Hash32)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
Please also add the SetLifetime calls for the LRUs (from the otel_reporter.go), otherwise the purgeTick code won't work.